core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -1069,6 +1119,101 @@ private void reduceCasts(RexCall outerCast) {
       }
     }
   }
+
+  /**
+   * Generate a new set of reduction options for controlling rule construction.
+   *
+   * @return A new set of reduction options.
+   */
+  public static ReductionOptions newOptions() {
+    return new ReductionOptions();
+  }
+
+  /**
+   * An immutable object capturing the set of options to be used for reducing
+   * expressions.
+   */
+  public static class ReductionOptions {
+    protected final boolean matchNullability;
```

Comment:
if the object is public, maybe there should be a public way to access its content